src/ray/gcs/gcs_server/gcs_server.cc

```diff
@@ -64,6 +60,28 @@ void GcsServer::Start() {
       main_service_, redis_client_->GetPrimaryContext(), [this]() { Stop(); });
   gcs_redis_failure_detector_->Start();
 
+  // Init gcs table storage.
+  if (RayConfig::instance().gcs_storage() == "redis") {
+    gcs_table_storage_ = std::make_shared<gcs::RedisGcsTableStorage>(redis_client_);
+  } else if (RayConfig::instance().gcs_storage() == "memory") {
+    gcs_table_storage_ = std::make_shared<InMemoryGcsTableStorage>(main_service_);
+  } else {
+    RAY_LOG(FATAL) << "Unsupported gcs storage: " << RayConfig::instance().gcs_storage();
+  }
+  // Init internal config
+  auto on_done = [this](const ray::Status &status) {
+    RAY_CHECK(status.ok()) << "Failed to put internal config";
+    this->main_service_.stop();
+  };
+  ray::rpc::StoredConfig stored_config;
+  stored_config.set_config(config_.config_list);
+  RAY_CHECK_OK(gcs_table_storage_->InternalConfigTable().Put(ray::UniqueID::Nil(),
+                                                             stored_config, on_done));
+  // Run the main service to make the Put in sync mode
+  main_service_.run();
+  // Reset the main service to the initial status
+  main_service_.restart();
```

Comment:
Why do we need a restart()?

Reply:
That's because of signal catching in the [main function](https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_server/gcs_server_main.cc#L124-L137). 
If not restart, when we hit that part, it'll call the handler immediately